example_allocators/bitmap_allocator/bitmap_alloc.c

```diff
@@ -0,0 +1,202 @@
+/**********************************
+ * bitmap_alloc.c
+ * Jeremy.Singer@glasgow.ac.uk
+ *
+ * This is a simple fixed-size bitmap allocator.
+ * It mmaps a large buffer of
+ * NUM_CHUNKS * CHUNK_SIZE bytes
+ * then allocates this space in equally-sized
+ * chunks to client code.
+ * A side bitmap is required to keep track of which
+ * chunks are in use (corresponding bit set to 1)
+ * and which chunks are free (corresponding bit
+ * set to 0). There is one bit per allocatable chunk.
+ *
+ * This is _not_ a clever allocator, since it
+ * does a linear scan of the bitmap to find the
+ * first free chunk, which is expensive!
+ * More efficient scans could be easily incorporated.
+ *
+ * This is _not_ a general-purpose allocator, since
+ * it only allocates chunks of a fixed size. Further,
+ * this size is constrained to be small enough to allow
+ * exact bounds representation in CHERI capabilities.
+ *
+ * This is an initial simple memory allocator test
+ * for CHERI / Capable VMs.
+ * We explore capability alignment,
+ * representable bounds, narrowing operations
+ * and compiler intrinsic support.
+ */
+
+// #include <cheriintrin.h>
+#include <assert.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+
+#include "bitmap_alloc.h"
+
+#define BITS_PER_BYTE 8
+
+char *buffer = NULL;		  /* allocation buffer */
+unsigned char *bitmap = NULL; /* bitmap for the buffer */
+
+int buffer_size = 0;	 /* size of buffer (in bytes) */
+int bitmap_size = 0;	 /* size of bitmap (in bytes) */
+int bytes_per_chunk = 0; /* size of single chunk (in bytes) */
+
+void init_alloc(int num_chunks, int chunk_size)
+{
+	int i = 0;
+
+	/* we need to increase the num_chunks
+	 * so every bit in bitmap will be used
+	 */
+	int adjusted_num_chunks = (num_chunks % BITS_PER_BYTE == 0)
+								  ? num_chunks
```

Comment:
The large indentation here makes this very difficult to read.

Reply:
This is set by clang-format. Should we change clang-format in this PR or should this be done separately ?